### PR TITLE
[types/factory_girl] Refine initializer, attributes and buildOptions

### PR DESCRIPTION
**Pull request by ishikawa**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
  - Documentation https://github.com/simonexmachina/factory-girl

### CODE REVIEW

File: types/factory-girl/factory-girl-tests.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
-import * as factory from "factory-girl";
+import * as factory from 'factory-girl';
```
Comment (ishikawa) [inline types/factory-girl/factory-girl-tests.ts:1]: I applied Prettier in e97e0bc.

File: types/factory-girl/index.d.ts [hunk 2]
```diff
@@ -12,11 +12,20 @@ declare const factory: factory.Static;
 declare namespace factory {
     type Generator<T> = () => T;
 
-    type Definition<T> = T | Generator<T>;
+    type Definition<T> = T | Generator<T> | Promise<T>;
```
Comment (ishikawa) [inline types/factory-girl/index.d.ts:15]: Each attribute can be `Promise<T>`.

- https://github.com/simonexmachina/factory-girl/blob/f7ec9d5004021ff1334ab9440e9b5cc4fd1a18b0/src/Factory.js#L43
- https://github.com/simonexmachina/factory-girl/blob/f7ec9d5004021ff1334ab9440e9b5cc4fd1a18b0/src/utils/asyncPopulate.js#L23

File: types/factory-girl/index.d.ts [hunk 3]
```diff
@@ -12,11 +12,20 @@ declare const factory: factory.Static;
...
 
-    type Attributes<T> = Definition<{
+    type Attributes<T> = {
         [P in keyof T]: Definition<T[P]>;
-    }>;
+    };
+
+    type MaybeReadonlyArray<T> = T | ReadonlyArray<T>;
```
Comment (ishikawa) [inline types/factory-girl/index.d.ts:21]: Most of `Factory#...Many()` function can take `T` or `ReadonlyArray<T>` for its attributes and options. So I defined a utility type for it.

- https://github.com/simonexmachina/factory-girl/blob/f7ec9d5004021ff1334ab9440e9b5cc4fd1a18b0/src/Factory.js#L69

File: types/factory-girl/index.d.ts [hunk 4]
```diff
@@ -39,23 +48,32 @@ declare namespace factory {
...
          * Builds a new model instance that is not persisted
          */
-        build<T>(name: string, attrs?: Attributes<Partial<T>>): Promise<T>;
+        build<T>(name: string, attrs?: Attributes<Partial<T>>, buildOptions?: BuildOptions): Promise<T>;
 
         /**
          * Builds an array of model instances that are persisted
          */
-        buildMany<T>(name: string, num: number, attrs?: Attributes<Partial<T>>): Promise<T[]>;
-        buildMany<T>(name: string, attrs?: ReadonlyArray<Attributes<Partial<T>>>): Promise<T[]>;
```
Comment (ishikawa) [inline types/factory-girl/index.d.ts:58]: As I investigated the implementation, `Factory#buildMany(...)` needs `count` parameter.

- https://github.com/simonexmachina/factory-girl/blob/f7ec9d5004021ff1334ab9440e9b5cc4fd1a18b0/src/Factory.js#L99

File: types/factory-girl/index.d.ts [hunk 5]
```diff
@@ -65,18 +83,27 @@ declare namespace factory {
...
         /**
          * Builds an array of model instances that are persisted
          */
-        createMany<T>(name: string, num: number, attrs?: Attributes<Partial<T>>, buildOptions?: Options<T>): Promise<T[]>;
-        createMany<T>(name: string, attrs?: ReadonlyArray<Attributes<Partial<T>>>, buildOptions?: Options<T>): Promise<T[]>;
+        createMany<T>(
+            name: string,
+            num: number,
+            attrs?: MaybeReadonlyArray<Attributes<Partial<T>>>,
+            buildOptions?: MaybeReadonlyArray<BuildOptions>,
```
Comment (ishikawa) [inline types/factory-girl/index.d.ts:95]: A `Options<T>` option can't be specified here. You can specify these options in `Factory#define()` or `Factory#extend()`.

- https://github.com/simonexmachina/factory-girl/blob/f7ec9d5004021ff1334ab9440e9b5cc4fd1a18b0/src/Factory.js#L110

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## factory-girl ([<kbd>unpkg</kbd>](https://unpkg.com/browse/factory-girl@latest/))
was missing the following properties:
1. ObjectAdapter
1. BookshelfAdapter
1. DefaultAdapter
1. MongooseAdapter
1. SequelizeAdapter

<details>
<summary>as well as these 1 other properties...</summary>
<p>ReduxORMAdapter</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 30aa4563d38b96d01ea7955bb033fcf50dc17574
</p>

**Review by OlivierKamers — APPROVED** [review]
Looks good! I linked these definitions to our project and it fixes some @ts-expect-errors we had 🎉

**Comment by elcuy** [comment]
Ready to merge

File: types/factory-girl/index.d.ts [hunk 6]
```diff
@@ -39,23 +48,32 @@ declare namespace factory {
...
 
         /**
          * Generates and returns a collection of model attributes as an object hash instead of the model instance
          */
-        attrsMany<T>(name: string, num: number, attrs?: ReadonlyArray<Attributes<Partial<T>>>): Promise<T[]>;
+        attrsMany<T>(
+            name: string,
+            num: number,
+            attrs?: MaybeReadonlyArray<Attributes<Partial<T>>>,
+            buildOptions?: BuildOptions | ReadonlyArray<BuildOptions>,
```
[the remark this answers is not in the document]
Comment (ishikawa) [reply types/factory-girl/index.d.ts:60]: Thanks! I hadn't noticed that. But sorry, it has been already merged 🙇

**Pull request merged**